### PR TITLE
Syntax classification improvement

### PR DESCRIPTION
**Pull request by Cosifne**
Potentially improve the perf in this https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1716804/

When we use DFS to visit the node syntax tree, the stack might be large since all the syntax nodes might be stored in it. (From the sample dump, it has more than 8k in the array) 

The solution here is to store the enumerator in the stack (Thanks Cyrus's for the idea). For each node in the tree, we only need to put one enumerator onto the stack, rather than put all its children to it.

Perf:
Before:
image

After:

image

Note:
It's slightly slower in `SimpleEditAtTheBegining`, but faster in all other scenarios.

### CODE REVIEW

**Comment by ToddGrun** [comment]
This code feels a bit more complex than I would have thought necessary, and I'm unsure why it's written to avoid recursion. I took a quick crack at writing it with recursion [here](https://github.com/dotnet/roslyn/pull/70774), and things seemed simpler and faster. Thoughts?

**Comment by ToddGrun** [comment]
I noticed "the" mis-spelled in the original code, which means @CyrusNajmabadi wrote it and might have context on why it intentionally avoids recursion

---
In reply to: [1806891746](https://github.com/dotnet/roslyn/pull/70772#issuecomment-1806891746) [](http://example.com/codeflow?ancestors=1806891746)

File: src/Tools/IdeCoreBenchmarks/SyntacticChangeRangeBenchmark.cs [hunk 1]
```diff
@@ -66,6 +66,23 @@ private SyntaxNode WithDestabalizingEditAtMiddle()
...
 
+        private SyntaxNode WithSimpleEditAtBeginning()
+        {
+            var addedText = "using System.IO;";
+            var newText = _text.WithChanges(new TextChange(new TextSpan(0, addedText.Length), addedText));
+            var newTree = _tree.WithChangedText(newText);
+            var newRoot = newTree.GetRoot();
+            return newRoot;
+        }
+        private SyntaxNode WithSimpleEditInTheEnd()
```
Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/SyntacticChangeRangeBenchmark.cs:77]:
```suggestion

        private SyntaxNode WithSimpleEditInTheEnd()
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Added some tests :)

**Comment by CyrusNajmabadi** [comment]
> and I'm unsure why it's written to avoid recursion. I took a quick crack at writing it with recursion https://github.com/dotnet/roslyn/pull/70774, and things seemed simpler and faster. Thoughts?

We can easily blow the stack with recursion in normal syntax trees.  Specifically, the common pattern of `"x" + "y" + "z" + ...` ends up being an a linear tree, which can easily hit stack limits if we recurse.

If we were to do it over again, i'd make it so that we had a NAryExpression for this sort of construct.  But c'est la vie.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun since you're invested now, can you add a test with an extremely large binary expr add tree, where an edit is then made in the middle?  it should demonstrate the stack overflow for the recursive case, and then show we're fine being non-recursive.

**Comment by ToddGrun** [comment]
Dang, I don't know if I want to be invested. :) I'm uncertain if roslyn generally avoids recursve tree walking (are there really non-crafted cases where recursive tree walking actually blows the stack?). I'm certain I could blow the stack with some obtuse case with my changes, just not sure how roslyn prioritizes that (seemingly quite unlikely) scenario.

---
In reply to: [1807005469](https://github.com/dotnet/roslyn/pull/70772#issuecomment-1807005469) [](http://example.com/codeflow?ancestors=1807005469)

**Comment by CyrusNajmabadi** [comment]
> (are there really non-crafted cases where recursive tree walking actually blows the stack?)

Yup.  The case i mentioned is one of those.  In real world code (and often in generated), concatenating large expressions is very common.  And we almost always blow the stack doing a tree recursion in some feature that isn't iterative.

**Comment by ToddGrun** [comment]
> > (are there really non-crafted cases where recursive tree walking actually blows the stack?)
> 
> Yup. The case i mentioned is one of those. In real world code (and often in generated), concatenating large expressions is very common. And we almost always blow the stack doing a tree recursion in some feature that isn't iterative.

I was able to repro by crafting a file using deeply nested binary expressions. Are there structure walking helpers (similar to SyntaxWalker) that aren't recursion based? Aren't all SyntaxWalker derived classes in danger of blowing the stack?

**Comment by ToddGrun** [comment]
> @ToddGrun since you're invested now, can you add a test with an extremely large binary expr add tree, where an edit is then made in the middle? it should demonstrate the stack overflow for the recursive case, and then show we're fine being non-recursive.

Added a unit test that would blow the stack with the implementation that I tried. And this implementation passes it with flying colors.

**Comment by CyrusNajmabadi** [comment]
Thanks Todd!

**Pull request merged**